fix(sourceprocessor): fix templating for sourceCategoryPrefix

aboguszewski-sumo opened the pull request:
Fixes #1339 - please see the issue to see the reason why it was reopened.

It turns out that the previous fix was not enough, because the code was affected also in some other places. I also did a small refactor.

sumo-drosiek commented on pkg/processor/sourceprocessor/source_category_filler.go:
```diff
@@ -76,21 +75,30 @@ func (f *sourceCategoryFiller) fill(attributes *pcommon.Map) {
 	}
 	valueTemplate = prefix + valueTemplate
 
+	if doesUseAnnotation {
+		templateAttributes = extractTemplateAttributes(valueTemplate)
+	} else {
+		templateAttributes = f.templateAttributes
+	}
```
Is it expected that we skip prefix here?

aboguszewski-sumo replied:
`valueTemplate` already contains the prefix, so all templates in the prefix will be affected.

sumo-drosiek replied:
I was asking about `f.templateAttributes`. Does it contain correct prefix as well?

aboguszewski-sumo replied:
Yes, see line `41`

astencel-sumo commented on pkg/processor/sourceprocessor/source_category_filler.go:
```diff
@@ -76,21 +75,30 @@ func (f *sourceCategoryFiller) fill(attributes *pcommon.Map) {
 	}
 	valueTemplate = prefix + valueTemplate
 
+	if doesUseAnnotation {
```
Why does it matter if we use annotations or not? Should prefix templating only work when using annotations?

aboguszewski-sumo replied:
This is a good question to the author of the processor.
As I understand it: if we use the annotations, then we override the standard template value or the prefix, and then we can't use the preprocessed template attributes array.

astencel-sumo commented on pkg/processor/sourceprocessor/source_category_filler_test.go:
```diff
@@ -53,14 +53,15 @@ func TestFillWithAnnotations(t *testing.T) {
+	attrs.PutStr("k8s.pod.name", "ABC")
 	attrs.PutStr("k8s.pod.annotation.sumologic.com/sourceCategory", "sc-from-annot-%{k8s.namespace.name}-%{k8s.pod.uid}")
-	attrs.PutStr("k8s.pod.annotation.sumologic.com/sourceCategoryPrefix", "%{k8s.pod.uid}-Prefix:")
+	attrs.PutStr("k8s.pod.annotation.sumologic.com/sourceCategoryPrefix", "%{k8s.pod.name}-%{k8s.pod.uid}-Prefix:")
```
Why is prefix templating tested only with pod annotations? What about `TestFill` and `TestFillWithContainerAnnotations`?

aboguszewski-sumo replied:
While this shouldn't matter, it's a good point that it should have more tests. Fixed.

astencel-sumo reviewed (changes requested):
Let's get the questions answered before we merge this.

sumo-drosiek reviewed (approved):
LGTM
Waiting for @astencel-sumo

aboguszewski-sumo commented:
@astencel-sumo

astencel-sumo commented on pkg/processor/sourceprocessor/source_category_filler_test.go:
```diff
@@ -131,4 +134,20 @@ func TestFillWithContainerAnnotations(t *testing.T) {
 		assertAttribute(t, attrs, "_sourceCategory", "THIRD_s-c!")
 	})
+
+	t.Run("set source category prefix", func(t *testing.T) {
```
I'm not sure what is the intent of this test? It does not enable container annotations (`cfg.ContainerAnnotations.Enabled = true`), so it only tests that the source category configured in the processor's settings is used.

We should rather have a test that defines what happens if you try to use templates in per-container annotations. I propose something like this:

```go
	t.Run("container annotations do not support templating", func(t *testing.T) {
		cfg := createDefaultConfig().(*Config)
		cfg.ContainerAnnotations.Enabled = true
		cfg.SourceCategory = "my-source-category"
		cfg.SourceCategoryPrefix = "my-prefix"

		attrs := pcommon.NewMap()
		attrs.PutStr("k8s.container.name", "container-name-1")
		attrs.PutStr("k8s.namespace.name", "ns-1")
		attrs.PutStr("k8s.pod.name", "ABC")
		attrs.PutStr("k8s.pod.annotation.sumologic.com/sourceCategory", "sc-from-annot-%{k8s.namespace.name}-%{k8s.pod.uid}")
		attrs.PutStr("k8s.pod.annotation.sumologic.com/sourceCategoryPrefix", "%{k8s.pod.name}-%{k8s.pod.uid}-Prefix:")
		attrs.PutStr("k8s.pod.annotation.sumologic.com/sourceCategoryReplaceDash", "#")
		attrs.PutStr("k8s.pod.annotation.sumologic.com/container-name-1.sourceCategory", "%{k8s.namespace.name}-%{k8s.pod.name}")
		attrs.PutStr("k8s.pod.annotation.sumologic.com/container-name-2.sourceCategory", "another/source-category")
		attrs.PutStr("k8s.container.name", "container-name-1")

		filler := newSourceCategoryFiller(cfg, zap.NewNop())
		filler.fill(&attrs)

		assertAttribute(t, attrs, "_sourceCategory", "%{k8s.namespace.name}-%{k8s.pod.name}")
	})
```

astencel-sumo reviewed (approved):
Ideally let's fix that test, other than that 🚀

aboguszewski-sumo commented:
@astencel-sumo I think that I added that test because there was no test that just checks the prefixing (I had some problems while implementing this and this test helped me).
I just added your test, I guess it should be fine.
